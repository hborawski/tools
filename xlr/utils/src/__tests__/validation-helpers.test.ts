import { test, expect, describe } from "vitest";
import type { ObjectType } from "@player-tools/xlr";
import { computeEffectiveObject } from "../validation-helpers";

describe("computeEffectiveObject tests", () => {
  test("mixed test", () => {
    const type1: ObjectType = {
      type: "object",
      properties: {
        foo: {
          required: true,
          node: {
            type: "string",
          },
        },
      },
      additionalProperties: false,
    };

    const type2: ObjectType = {
      type: "object",
      properties: {
        bar: {
          required: true,
          node: {
            type: "number",
          },
        },
      },
      additionalProperties: {
        type: "unknown",
      },
    };

    expect(computeEffectiveObject(type1, type2)).toMatchSnapshot();
  });

  test("Error on property overlap", () => {
    const type1: ObjectType = {
      type: "object",
      properties: {
        foo: {
          required: true,
          node: {
            type: "string",
          },
        },
      },
      additionalProperties: false,
    };

    const type2: ObjectType = {
      type: "object",
      properties: {
        foo: {
          required: true,
          node: {
            type: "number",
          },
        },
      },
      additionalProperties: {
        type: "unknown",
      },
    };

    expect(() =>
      computeEffectiveObject(type1, type2, true)
    ).toThrowErrorMatchingInlineSnapshot(
      `[Error: Can't compute effective type for object literal and object literal because of conflicting properties foo]`
    );
  });

  test("Merges equal additionalProperties", () => {
    const type1: ObjectType = {
      type: "object",
      properties: {
        foo: {
          required: true,
          node: {
            type: "string",
          },
        },
      },
      additionalProperties: {
        type: "unknown",
      },
    };

    const type2: ObjectType = {
      type: "object",
      properties: {
        bar: {
          required: true,
          node: {
            type: "number",
          },
        },
      },
      additionalProperties: {
        type: "unknown",
      },
    };

    expect(computeEffectiveObject(type1, type2)).toMatchSnapshot();
  });
});
