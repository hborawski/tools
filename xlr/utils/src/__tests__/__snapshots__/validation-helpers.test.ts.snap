// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`computeEffectiveObject tests Merges equal additionalProperties 1`] = `
Object {
  "additionalProperties": Object {
    "type": "unknown",
  },
  "description": "Effective type combining object literal and object literal",
  "genericTokens": Array [],
  "name": "object literal & object literal",
  "properties": Object {
    "bar": Object {
      "node": Object {
        "type": "number",
      },
      "required": true,
    },
    "foo": Object {
      "node": Object {
        "type": "string",
      },
      "required": true,
    },
  },
  "type": "object",
}
`;

exports[`computeEffectiveObject tests mixed test 1`] = `
Object {
  "additionalProperties": Object {
    "type": "unknown",
  },
  "description": "Effective type combining object literal and object literal",
  "genericTokens": [],
  "name": "object literal & object literal",
  "properties": {
    "bar": {
      "node": {
        "type": "number",
      },
      "required": true,
    },
    "foo": {
      "node": {
        "type": "string",
      },
      "required": true,
    },
  },
  "type": "object",
}
`;
